### PR TITLE
whisper using player's mojang provided public key to encrypt

### PR DESCRIPTION
**Pull request by wagyourtail**
this is bad xd

### CODE REVIEW

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+import java.util.Optional;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+import static com.mojang.brigadier.arguments.StringArgumentType.string;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.argument;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.literal;
+import static net.minecraft.command.CommandSource.suggestMatching;
+
+public class WhisperEncrypted {
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Could you add the `Command` suffix to this class?

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import java.security.PublicKey;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+import static com.mojang.brigadier.arguments.StringArgumentType.string;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.argument;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.literal;
+import static net.minecraft.command.CommandSource.suggestMatching;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: `import static`s should use wildcard imports.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+public class WhisperEncrypted {
+    private static boolean justSent = false;
+
+    public static void register(CommandDispatcher<FabricClientCommandSource> dispatcher) {
+        dispatcher.register(literal("cwhisperencrypted")
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: I think this root literal is rather long, what about `/cwe`?

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                .then(argument("player", string())
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: You can use the `CGameProfileArgumentType`.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                        .suggests((context, builder) -> suggestMatching(context.getSource().getPlayerNames(), builder))
+                        .then(argument("message", StringArgumentType.greedyString())
+                            .executes(WhisperEncrypted::whisper))));
+    }
+
+    private static int whisper(CommandContext<FabricClientCommandSource> context) {
+        String message = context.getArgument("message", String.class);
+        String player = context.getArgument("player", String.class);
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: It's usual to add the `FabricClientCommandSource` and the command parameters as parameters to the method. You can also use the `FabricClientCommandSource` to access the `MinecraftClient` instance.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        Optional<PlayerListEntry> entry = MinecraftClient.getInstance().getNetworkHandler().getPlayerList().stream().filter(e -> e.getProfile().getName().equalsIgnoreCase(player)).findFirst();
+        if (entry.isEmpty()) {
+            context.getSource().sendError(Text.translatable("commands.cwhisperencrypted.player_not_found"));
+            return 0;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: It's usual to throw en exception when the command fails somehow.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        }
+        try {
+            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
+            // step 2 encrypt the encrypted message using the public key of the recipient
+            PlayerPublicKey ppk = entry.get().getPublicKeyData();
+            if (ppk == null) {
+                context.getSource().sendError(Text.translatable("commands.cwhisperencrypted.pub_key_not_found"));
+                return 0;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+            }
+            PublicKey publicKey = ppk.data().key();
+            cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
+            cipher.init(Cipher.ENCRYPT_MODE, publicKey);
+            byte[] encrypted = message.getBytes(StandardCharsets.UTF_8);
+            encrypted = Gzip.compress(encrypted);
+            if (encrypted.length > 245) {
+                context.getSource().sendError(Text.translatable("commands.cwhisperencrypted.message_too_long"));
+                return 0;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+            }
+            cipher.update(encrypted);
+            encrypted = cipher.doFinal();
+            String commandMessage = "w " + player + " CCENC:" + BaseUTF8.toUnicode(encrypted);
+            if (commandMessage.length() >= 256) {
+                context.getSource().sendError(Text.translatable("commands.cwhisperencrypted.message_too_long"));
+                return 0;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+            }
+            justSent = true;
+            MinecraftClient.getInstance().player.sendCommand(commandMessage);
+        } catch (Exception e) {
+            context.getSource().sendError(Text.translatable("commands.cplayerinfo.ioException"));
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above and perhaps you could create a custom error message for this command.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+            e.printStackTrace();
+            return 0;
+        }
+        return 1;
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Return `Command.SINGLE_SUCCESS` instead.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+            if (str == null || str.length == 0) {
+                return null;
+            }
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            GZIPOutputStream gzip;
+            try {
+                gzip = new GZIPOutputStream(out);
+                gzip.write(str);
+                gzip.close();
+            } catch ( Exception e) {
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]:
```suggestion
            } catch (Exception e) {
```

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+        public static String toUnicode(byte[] b) {
+            StringBuilder sb = new StringBuilder();
+            int data = 0;
+            int bitPtr = 0;
+            for (byte b1 : b) {
+                data |= Byte.toUnsignedInt(b1) << bitPtr;
+                bitPtr += 8;
+                if (bitPtr > 19) {
+                    int codePoint = mapCodepoint((data & 0x0FFFFF) + LOWER_BOUND);
+                    if (codePoint >= UPPER_BOUND) throw new RuntimeException("WHAT?");
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Every `if`-statement should have curly brackets.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+                    data >>= 20;
+                    sb.appendCodePoint(codePoint);
+                }
+                bitPtr %= 20;
+            }
+            if (bitPtr != 0) {
+                int codePoint = mapCodepoint((data & 0x0FFFFF) + LOWER_BOUND);
+                if (codePoint >= UPPER_BOUND) throw new RuntimeException("WHAT?");
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+        public static byte[] fromUnicode(String s) {
+            int bitLength = s.codePointCount(0, s.length()) * 20;
+            int dataLength = (bitLength + 7) / 8;
+            byte[] data = new byte[dataLength];
+            int dataPtr = 0;
+            for (int i = 0; i < s.length(); ) {
+                int codepoint = s.codePointAt(i);
+                i += Character.charCount(codepoint);
+                if (codepoint >= UPPER_BOUND) throw new RuntimeException("WHAT?");
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above.

File: src/main/resources/assets/clientcommands/lang/en_us.json [hunk 16]
```diff
@@ -176,6 +176,12 @@
 
   "commands.cwiki.failed": "Could not retrieve wiki content",
 
+  "commands.cwhisperencrypted.message_too_long": "Message too long",
+  "commands.cwhisperencrypted.player_not_found": "Player not found",
+  "commands.cwhisperencrypted.no_priv_key": "No private key found",
+  "commands.cwhisperencrypted.pub_key_not_found": "Public key not found",
```
Comment (xpple) [inline src/main/resources/assets/clientcommands/lang/en_us.json]: This doesn't match the convention of using camelCase for translation keys.

File: src/main/resources/assets/clientcommands/lang/en_us.json [hunk 17]
```diff
@@ -176,6 +176,12 @@
...
+
+
```
Comment (xpple) [inline src/main/resources/assets/clientcommands/lang/en_us.json:207]:
```suggestion
```

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+            }
+            return sb.toString();
+        }
+
+        public static byte[] fromUnicode(String s) {
+            int bitLength = s.codePointCount(0, s.length()) * 20;
+            int dataLength = (bitLength + 7) / 8;
+            byte[] data = new byte[dataLength];
+            int dataPtr = 0;
+            for (int i = 0; i < s.length(); ) {
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]:
```suggestion
            for (int i = 0; i < s.length();) {
```

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+                    }
+                    cipher.init(Cipher.DECRYPT_MODE, key.get());
+                    encrypted = cipher.doFinal(encrypted);
+                    encrypted = Gzip.uncompress(encrypted);
+                    String message = "ccenc: " + new String(encrypted, StandardCharsets.UTF_8);
+                    return new JsonPrimitive(message);
+                } catch (Exception ex) {
+                    if (!justSent) {
+                        MinecraftClient.getInstance().inGameHud.getChatHud().addMessage(Text.translatable(
+                            "commands.cplayerinfo.ioException").formatted(Formatting.DARK_RED));
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncrypted.java]: Same as above, regarding the custom error message.

**Review by xpple — CHANGES_REQUESTED** [review]
Overall really cool command, just a few code style suggestions.

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncryptedCommand.java [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+import static dev.xpple.clientarguments.arguments.CGameProfileArgumentType.*;
+import static net.fabricmc.fabric.api.client.command.v2.ClientCommandManager.*;
+
+public class WhisperEncryptedCommand {
+    private static boolean justSent = false;
+
+    public static void register(CommandDispatcher<FabricClientCommandSource> dispatcher) {
+        dispatcher.register(literal("cwe")
+                .then(argument("player", string())
+                    .then(argument("message", gameProfile())
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncryptedCommand.java]:
```suggestion
                .then(argument("player", gameProfile())
                    .then(argument("message", string())
```

File: src/main/java/net/earthcomputer/clientcommands/command/WhisperEncryptedCommand.java [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+                        .executes((ctx) ->
+                            WhisperEncryptedCommand.whisper(ctx,
+                                CGameProfileArgumentType.getCProfileArgument(ctx, "player"),
+                                ctx.getArgument("message", String.class)
+                            )
```
Comment (xpple) [inline src/main/java/net/earthcomputer/clientcommands/command/WhisperEncryptedCommand.java]:
```suggestion
                            whisper(ctx.getSource(),
                                CGameProfileArgumentType.getCProfileArgument(ctx, "player"),
                                getString(ctx, "message")
                            )
```

**Review by xpple — CHANGES_REQUESTED** [review]

**Review by xpple — APPROVED** [review]

**Comment by Earthcomputer** [comment]
Superseded by #461

**Pull request closed without merging**